proposals/safe-fixed-sized-buffers.md

```diff
@@ -0,0 +1,121 @@
+    ;
+    
+fixed_size_buffer_declarator
+    : identifier '[' constant_expression ']'
+    ;    
+```
+
+A *fixed_size_buffer_declarator* introduces a fixed-size buffers of a given element type.
+
+The buffer element type is the *type* specified in `field_declaration`. A fixed-size buffer declarator introduces a new member and consists of an identifier that names the member, followed by a constant expression enclosed in `[` and `]` tokens. The constant expression denotes the number of elements in the member introduced by that fixed-size buffer declarator. The type of the constant expression must be implicitly convertible to type `int`, and the value must be a non-zero positive integer.
+
+The elements of a fixed-size buffer shall be laid out sequentially in memory as though they are elements of an array.
+
+A *field_declaration* with a *fixed_size_buffer_declarator* cannot have the `volatile` modifier, and it cannot be marked with a `System.ThreadStaticAttribute`.
+A *field_declaration* with a *fixed_size_buffer_declarator* in an interface must have `static` modifier.
+
+Depending on the situation (details are specified below), an access to a fixed-size buffer member is classified as a value (never a variable) of either
+`System.ReadOnlySpan<S>` or `System.Span<S>`, where S is the element type of the fixed-size buffer. Both types provide indexers returning a reference to a
+specific element with appropriate "readonly-ness", which prevents direct assignment to the elements when language rules don't permit that.
+
+The resulting span instance will have a length equal to the size declared on the fixed-size buffer.
+Indexing into the span with a constant expression outside of the declared fixed-size buffer bounds is a compile time error.
+
+The *safe-to-escape* scope of the value will be equal to the *safe-to-escape* scope of the container, just as it would if the backing data was accessed as a field.
+
+### Fixed-size buffers in expressions
+
+Member lookup of a fixed-size buffer member proceeds exactly like member lookup of a field.
+
+A fixed-size buffer can be referenced in an expression using a *simple_name* or a *member_access* .
+
+When an instance fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `this.I`, where `I` is the fixed-size buffer member. When a static fixed-size buffer member is referenced as a simple name, the effect is the same as a member access of the form `E.I`, where `I` is the fixed-size buffer member and `E` is the declaring type.
+
+#### Non-readonly fixed-size buffers
+
+In a member access of the form `E.I`, if `E` is of a struct type and a member lookup of `I` in that struct type identifies an instance fixed-size member,
+then `E.I` is evaluated and classified as follows:
+
+- If `E` is classified as a value, the result of the expression is classified as a value of type `System.ReadOnlySpan<S>`, where S is the element type of `I`.
```

Comment:
>If `E` is classified as a value [](http://example.com/codeflow?start=2&length=31)

Does this cover the case of a non-readonly fixed-size field in a `readonly` method?

Reply:
> Does this cover the case of a non-readonly fixed-size field in a `readonly` method?

I think it does, `this` is qualified as a value rather than a variable in a `readonly` method, I think. In any case, my intent is to cover this scenario as well. If I discover that a special spec wording is necessary for that, I will add it.